In [1]:
import pandas as pd
import numpy as np

print("Pandas:", pd.__version__)
print("NumPy:", np.__version__)

Pandas: 3.0.5
NumPy: 2.5.3


In [7]:
df = pd.read_csv("animal_data_dirty1.csv")

In [9]:
df.shape

(1011, 1)

In [10]:
df.columns.tolist()

['Animal type;Country;Weight kg;Body Length cm;Gender;Animal code;Latitude;Longitude;Animal name;Observation date;Data compiled by']

In [11]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1011 entries, 0 to 1010
Data columns (total 1 columns):
 #   Column                                                                                                                            Non-Null Count  Dtype
---  ------                                                                                                                            --------------  -----
 0   Animal type;Country;Weight kg;Body Length cm;Gender;Animal code;Latitude;Longitude;Animal name;Observation date;Data compiled by  1011 non-null   str  
dtypes: str(1)
memory usage: 8.0 KB


In [12]:
df.describe(include="all")

,Animal type;Country;Weight kg;Body Length cm;Gender;Animal code;Latitude;Longitude;Animal name;Observation date;Data compiled by
count,1011
unique,844
top,red squirel;Poland;0.308;22;female;;52.241273;...
freq,2


In [13]:
df.isnull().sum()

Animal type;Country;Weight kg;Body Length cm;Gender;Animal code;Latitude;Longitude;Animal name;Observation date;Data compiled by    0
dtype: int64

In [14]:
df.duplicated().sum()

np.int64(167)

In [16]:
df = pd.read_csv("animal_data_dirty1.csv", sep=";")
df.head()

,Animal type,Country,Weight kg,Body Length cm,Gender,Animal code,Latitude,Longitude,Animal name,Observation date,Data compiled by
0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,03.01.2024,James Johnson
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,03.02.2024,James Johnson
2,European bison,Poland,930.0,335.0,male,NaN,52.828845,23.820144,Szefu,01.03.2024,Anne Anthony
3,European bison,Poland,909.0,311.0,not determined,NaN,52.830509,23.826849,NaN,01.03.2024,Anne Anthony
4,European bison™,Poland,581.0,277.0,female,NaN,52.834109,23.807093,NaN,01.03.2024,Anne Anthony


In [17]:
df.shape

(1011, 11)

In [18]:
df.columns.tolist()

['Animal type',
 'Country',
 'Weight kg',
 'Body Length cm',
 'Gender',
 'Animal code',
 'Latitude',
 'Longitude',
 'Animal name',
 'Observation date',
 'Data compiled by']

In [19]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1011 entries, 0 to 1010
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Animal type       991 non-null    str    
 1   Country           999 non-null    str    
 2   Weight kg         984 non-null    float64
 3   Body Length cm    984 non-null    float64
 4   Gender            992 non-null    str    
 5   Animal code       0 non-null      float64
 6   Latitude          913 non-null    float64
 7   Longitude         913 non-null    float64
 8   Animal name       52 non-null     str    
 9   Observation date  1011 non-null   str    
 10  Data compiled by  1011 non-null   str    
dtypes: float64(5), str(6)
memory usage: 87.0 KB


In [22]:
df.describe(include="all")

,Animal type,Country,Weight kg,Body Length cm,Gender,Animal code,Latitude,Longitude,Animal name,Observation date,Data compiled by
count,991,999,984.000000,984.000000,992,0.0,913.000000,913.000000,52,1011,1011
unique,12,14,NaN,NaN,3,NaN,NaN,NaN,10,114,4
top,red squirrel,Poland,NaN,NaN,male,NaN,NaN,NaN,Bob Bobson,01.03.2024,Anne Anthony
freq,543,291,NaN,NaN,499,NaN,NaN,NaN,18,61,473
mean,NaN,NaN,39.745503,39.107724,NaN,NaN,49.393369,18.203280,NaN,NaN,NaN
std,NaN,NaN,156.290076,58.628601,NaN,NaN,7.168900,3.899601,NaN,NaN,NaN
min,NaN,NaN,-0.252000,-19.000000,NaN,NaN,-78.582973,11.074008,NaN,NaN,NaN
25%,NaN,NaN,0.293000,19.000000,NaN,NaN,48.186913,14.384559,NaN,NaN,NaN
50%,NaN,NaN,0.331500,21.000000,NaN,NaN,49.560723,18.944015,NaN,NaN,NaN
75%,NaN,NaN,0.800000,23.000000,NaN,NaN,52.212433,21.033243,NaN,NaN,NaN


In [20]:
df.isnull().sum()

Animal type           20
Country               12
Weight kg             27
Body Length cm        27
Gender                19
Animal code         1011
Latitude              98
Longitude             98
Animal name          959
Observation date       0
Data compiled by       0
dtype: int64

In [21]:
df.duplicated().sum()

np.int64(167)

In [23]:
print("Animal type:")
print(df["Animal type"].value_counts(dropna=False))

print("\nCountry:")
print(df["Country"].value_counts(dropna=False))

print("\nGender:")
print(df["Gender"].value_counts(dropna=False))

Animal type:
Animal type
red squirrel       543
hedgehog           274
lynx                61
European bison      49
red squirrell       22
NaN                 20
red squirel         14
lynx?               10
European bison™      6
European bisson      4
European buster      4
ledgehod             3
wedgehod             1
Name: count, dtype: int64

Country:
Country
Poland            291
Germany           176
Slovakia          151
Hungary           146
Czech Republic    103
Austria            74
PL                 15
NaN                12
HU                 11
Hungry             10
CZ                  5
DE                  5
Czech               4
Australia           4
CC                  4
Name: count, dtype: int64

Gender:
Gender
male              499
female            488
NaN                19
not determined      5
Name: count, dtype: int64


In [24]:
df[["Weight kg", "Body Length cm", "Latitude", "Longitude"]].describe()

,Weight kg,Body Length cm,Latitude,Longitude
count,984.000000,984.000000,913.000000,913.000000
mean,39.745503,39.107724,49.393369,18.203280
std,156.290076,58.628601,7.168900,3.899601
min,-0.252000,-19.000000,-78.582973,11.074008
25%,0.293000,19.000000,48.186913,14.384559
50%,0.331500,21.000000,49.560723,18.944015
75%,0.800000,23.000000,52.212433,21.033243
max,1100.000000,350.000000,52.853843,34.896734


In [25]:
print(df["Observation date"].head(10))
print("\nTipo de dato:")
print(df["Observation date"].dtype)

0    03.01.2024
1    03.02.2024
2    01.03.2024
3    01.03.2024
4    01.03.2024
5    01.03.2024
6    01.03.2024
7    01.03.2024
8    01.03.2024
9    01.03.2024
Name: Observation date, dtype: str

Tipo de dato:
str


In [26]:
df[df.duplicated(keep=False)].head(20)

,Animal type,Country,Weight kg,Body Length cm,Gender,Animal code,Latitude,Longitude,Animal name,Observation date,Data compiled by
9,red squirel,Poland,0.308,22.0,female,NaN,52.241273,21.055745,Basia,01.03.2024,Anne Anthony
10,red squirrel,PL,0.348,20.0,female,NaN,52.242060,21.057802,NaN,01.03.2024,Anne Anthony
11,red squirrel,Poland,0.316,20.0,male,NaN,52.240035,21.056188,Lolek,01.03.2024,Anne Anthony
12,red squirrel,Poland,0.340,21.0,NaN,NaN,52.240284,21.051948,NaN,01.03.2024,Anne Anthony
13,red squirrel,Poland,0.333,21.0,male,NaN,NaN,NaN,NaN,01.03.2024,Anne Anthony
14,red squirrel,Poland,0.265,19.0,female,NaN,52.240560,21.055899,NaN,01.03.2024,Anne Anthony
15,red squirrel,Poland,NaN,NaN,male,NaN,52.241177,21.052873,NaN,01.03.2024,Anne Anthony
16,red squirrel,PL,0.261,22.0,female,NaN,52.240833,21.052093,NaN,01.03.2024,Anne Anthony
17,red squirrel,Poland,0.317,23.0,male,NaN,52.241358,21.056352,NaN,01.03.2024,Anne Anthony
18,red squirrel,Poland,0.328,21.0,male,NaN,52.241905,21.050727,NaN,01.03.2024,Anne Anthony


In [27]:
# Guardamos indicadores del estado inicial

duplicados_antes = df.duplicated().sum()

pesos_negativos_antes = (df["Weight kg"] < 0).sum()
longitudes_negativas_antes = (df["Body Length cm"] < 0).sum()

print("Duplicados antes:", duplicados_antes)
print("Pesos negativos antes:", pesos_negativos_antes)
print("Longitudes negativas antes:", longitudes_negativas_antes)
print("Tipo de fecha antes:", df["Observation date"].dtype)

Duplicados antes: 167
Pesos negativos antes: 5
Longitudes negativas antes: 6
Tipo de fecha antes: str


In [28]:
animal_type_map = {
    "red squirrel": "red squirrel",
    "red squirrell": "red squirrel",
    "red squirel": "red squirrel",
    "lynx": "lynx",
    "lynx?": "lynx",
    "European bison": "European bison",
    "European bison™": "European bison",
    "European bisson": "European bison",
    "European buster": "European bison",
    "hedgehog": "hedgehog",
    "ledgehod": "hedgehog",
    "wedgehod": "hedgehog"
}

df["Animal type"] = df["Animal type"].replace(animal_type_map)

In [29]:
df["Country"] = df["Country"].replace({
    "PL": "Poland"
})

In [30]:
df["Gender"] = df["Gender"].fillna("not determined")

In [31]:
df["Observation date"] = pd.to_datetime(
    df["Observation date"],
    format="%d.%m.%Y",
    errors="coerce"
)

In [32]:
df.loc[df["Weight kg"] < 0, "Weight kg"] = np.nan

df.loc[df["Body Length cm"] < 0, "Body Length cm"] = np.nan

In [33]:
df = df.drop_duplicates()

In [34]:
print("========== VALIDACIÓN DE LA LIMPIEZA ==========")

print("\n1. Duplicados:")
print("Antes:", duplicados_antes)
print("Después:", df.duplicated().sum())

print("\n2. Valores negativos:")
print("Pesos negativos después:", (df["Weight kg"] < 0).sum())
print("Longitudes negativas después:", (df["Body Length cm"] < 0).sum())

print("\n3. Tipo de fecha:")
print(df["Observation date"].dtype)

print("\n4. Tipos de animal:")
print(df["Animal type"].value_counts(dropna=False))

print("\n5. Países:")
print(df["Country"].value_counts(dropna=False))

print("\n6. Género:")
print(df["Gender"].value_counts(dropna=False))

========== VALIDACIÓN DE LA LIMPIEZA ==========

1. Duplicados:
Antes: 167
Después: 0

2. Valores negativos:
Pesos negativos después: 0
Longitudes negativas después: 0

3. Tipo de fecha:
datetime64[us]

4. Tipos de animal:
Animal type
red squirrel      413
hedgehog          278
lynx               71
European bison     63
NaN                17
Name: count, dtype: int64

5. Países:
Country
Poland            192
Germany           173
Slovakia          151
Hungary           103
Czech Republic    103
Austria            74
NaN                 9
HU                  8
Hungry              7
CZ                  5
DE                  5
Czech               4
Australia           4
CC                  4
Name: count, dtype: int64

6. Género:
Gender
male              423
female            401
not determined     18
Name: count, dtype: int64


In [37]:
df[df["Observation date"].isna()][
    ["Animal type", "Country", "Weight kg", "Body Length cm", "Gender"]
]

,Animal type,Country,Weight kg,Body Length cm,Gender
968,NaN,NaN,NaN,NaN,not determined
972,hedgehog,Hungary,NaN,NaN,male
973,hedgehog,Hungary,0.900,16.0,female
974,red squirrel,Poland,0.281,21.0,male
975,hedgehog,Slovakia,0.600,19.0,female
976,lynx,Slovakia,27.000,96.0,male
978,lynx,Poland,13.000,63.0,female
979,lynx,Poland,20.000,87.0,female
980,red squirrel,Poland,0.282,22.0,female
981,red squirrel,Germany,0.305,21.0,female


In [38]:
df["Animal type"].value_counts(dropna=False)

Animal type
red squirrel      413
hedgehog          278
lynx               71
European bison     63
NaN                17
Name: count, dtype: int64

In [39]:
print("Duplicados después de la limpieza:", df.duplicated().sum())

Duplicados después de la limpieza: 0


Justificación de la limpieza: En esta etapa revisé y limpié los datos para corregir algunos errores que podían afectar el análisis. Se corrigieron nombres de animales y países que estaban escritos de diferentes formas, se trataron los datos faltantes de género, se convirtió la fecha a un formato adecuado y se corrigieron valores negativos que no tenían sentido para el peso y la longitud de los animales. También se eliminaron los registros duplicados para evitar contar la misma observación más de una vez. Los datos faltantes de algunas variables se mantuvieron cuando no era posible completar la información de forma confiable.

Punto 3. Crear una variable con np.where() "Variable longitud del Animal"

In [43]:
mediana_longitud = np.nanmedian(df["Body Length cm"])

print("Mediana de longitud corporal:", mediana_longitud, "cm")

Mediana de longitud corporal: 21.0 cm


In [44]:
df["Longitud corporal"] = np.where(
    df["Body Length cm"] >= mediana_longitud,
    "Igual o mayor a la mediana",
    "Menor a la mediana"
)

In [45]:
df[[
    "Animal type",
    "Body Length cm",
    "Longitud corporal"
]].head(15)

,Animal type,Body Length cm,Longitud corporal
0,NaN,NaN,Menor a la mediana
1,NaN,NaN,Menor a la mediana
2,European bison,335.0,Igual o mayor a la mediana
3,European bison,311.0,Igual o mayor a la mediana
4,European bison,277.0,Igual o mayor a la mediana
5,European bison,295.0,Igual o mayor a la mediana
6,European bison,250.0,Igual o mayor a la mediana
7,lynx,76.0,Igual o mayor a la mediana
8,lynx,82.0,Igual o mayor a la mediana
9,red squirrel,22.0,Igual o mayor a la mediana


In [46]:
df["Longitud corporal"].value_counts()

Longitud corporal
Igual o mayor a la mediana    451
Menor a la mediana            391
Name: count, dtype: int64

Función Estadistica, peso promedio del animal 

In [48]:
promedio_peso = np.nanmean(df["Weight kg"])

print("Peso promedio de los animales:", round(promedio_peso, 2), "kg")

Peso promedio de los animales: 47.58 kg


In [49]:
mediana_peso = np.nanmedian(df["Weight kg"])

print("Promedio de peso:", round(promedio_peso, 2), "kg")
print("Mediana de peso:", round(mediana_peso, 2), "kg")

Promedio de peso: 47.58 kg
Mediana de peso: 0.35 kg


La mediana de la longitud corporal fue de 21 cm, por lo que se utilizó este valor como referencia para crear una nueva variable mediante np.where(), clasificando las observaciones entre animales con una longitud igual o mayor a la mediana y animales con una longitud menor. Además, utilizando np.nanmean() se obtuvo un peso promedio de aproximadamente 47,58 kg. Al compararlo con la mediana de peso de 0,35 kg, se observa una diferencia considerable, posiblemente relacionada con la presencia de animales de diferentes tamaños y pesos y con valores extremos en la variable.

4. Cuántos registros existen para cada tipo de animal.

In [50]:
observaciones_animal = df.groupby("Animal type").size().sort_values(ascending=False)

print(observaciones_animal)

Animal type
red squirrel      413
hedgehog          278
lynx               71
European bison     63
dtype: int64


In [51]:
promedio_peso_animal = (
    df.groupby("Animal type")["Weight kg"]
    .mean()
    .sort_values(ascending=False)
)

print(promedio_peso_animal)

Animal type
European bison    588.396825
lynx               23.267606
hedgehog            0.797455
red squirrel        0.297567
Name: Weight kg, dtype: float64


¿Cuál es el promedio de una variable numérica según el país?

In [52]:
promedio_peso_pais = (
    df.groupby("Country")["Weight kg"]
    .mean()
    .sort_values(ascending=False)
)

print(promedio_peso_pais)

Country
CC                634.000000
CZ                150.535400
Poland            133.463989
Czech             100.908750
Czech Republic     48.050932
Slovakia           32.663740
Hungry              5.880333
HU                  3.563125
Hungary             2.786980
Austria             1.944041
DE                  0.636400
Australia           0.494000
Germany             0.450879
Name: Weight kg, dtype: float64


cantidad de animales por país

In [53]:
observaciones_pais = (
    df.groupby("Country")
    .size()
    .sort_values(ascending=False)
)

print(observaciones_pais)

Country
Poland            192
Germany           173
Slovakia          151
Czech Republic    103
Hungary           103
Austria            74
HU                  8
Hungry              7
CZ                  5
DE                  5
Czech               4
Australia           4
CC                  4
dtype: int64


En este punto utilicé groupby() para organizar la información y poder comparar los datos de acuerdo con el tipo de animal y el país. Se analizaron cuatro aspectos: la cantidad de observaciones por tipo de animal, el peso promedio por tipo de animal, el peso promedio por país y la cantidad de observaciones por país. Estos análisis permiten identificar qué animales tienen mayor presencia en los datos, comparar sus pesos promedio y observar cómo se distribuyen las observaciones entre los diferentes países. Esto facilita la interpretación de los datos y permite encontrar diferencias y patrones que no serían tan fáciles de identificar revisando los registros individualmente.

5. Conclusiones

1. El tipo de animal con mayor cantidad de observaciones es el red squirrel. La categoría red squirrel registra 413 observaciones, seguida por hedgehog con 278, lynx con 71 y European bison con 63. Esto indica que el red squirrel tiene la mayor representación dentro del conjunto de datos y, por lo tanto, tiene mayor peso en los análisis generales.

2. El European bison presenta el mayor peso promedio entre los tipos de animales analizados. El peso promedio del European bison es de aproximadamente 588,40 kg, mientras que el lynx presenta un promedio de 23,27 kg, el hedgehog de 0,80 kg y el red squirrel de 0,30 kg. Esto muestra una diferencia muy marcada en el tamaño y peso de las especies incluidas en el conjunto de datos.

3. Polonia es el país con mayor cantidad de observaciones.
Polonia presenta 192 observaciones, seguida por Alemania con 173 y Eslovaquia con 151. Esto significa que Polonia tiene la mayor representación de registros dentro del conjunto de datos, por lo que sus resultados tienen una participación importante en el análisis general.

4. Existen diferencias importantes en el peso promedio según el país y también inconsistencias en la forma de registrar los países.
El mayor promedio corresponde a CC, con 634 kg, seguido por CZ con aproximadamente 150,54 kg y Polonia con 133,46 kg. Sin embargo, se observa que algunos países aparecen con nombres completos y otros con códigos, por ejemplo Germany y DE, o Hungary y HU. Esto indica que la variable Country todavía presenta diferencias en la forma de registro, situación que debe tenerse en cuenta al interpretar los resultados.